docs/tuf-artifact-replication.adoc

```diff
@@ -0,0 +1,185 @@
+  artifacts (OS and zone images, firmware, etc.) even if it did store
+  them.
+* Sled Agent is generally what will directly use the artifacts (except
+  for SP and ROT images, which MGS needs), and it can also manage its
+  own local storage.
+
+Thus Nexus needs to accept artifacts from outside of the system and
+immediately offload them to individual Sled Agents for persistent
+storage and later use.
+
+We have chosen (see <<rfd424>>) the simplest possible implementation:
+every Sled Agent stores a copy of every artifact on each of its M.2
+devices. This is storage inefficient but means that a Sled Agent can
+directly use those resources to create zones from updated images,
+install an updated OS, or manage the installation of updates on other
+components, without Nexus having to ensure that it tells a sled to
+have an artifact _before_ telling it to use it. A Nexus background task
+periodically ensures that all sleds have all artifacts.
+
+== Sled Agent implementation
+
+Sled Agent stores artifacts as a content-addressed store on an *update*
+dataset on each M.2 device: the file name of each stored artifact is its
+SHA-256 hash.
+
+It also stores an _artifact configuration_ in memory: a list of all
+artifact hashes that the sled should store, and a generation number.
+The generation number is owned by Nexus, which increments the generation
+number when the set of TUF repos on the system changes. Sled Agent
+prevents modifying the configuration without an increase in the
+generation number.
+
+Sled Agent offers the following APIs on the underlay network, intended
+for Nexus:
+
+* `artifact_config_get`: Get the current artifact configuration.
+* `artifact_config_put`: Put the artifact configuration that should be
+  in effect. This API is idempotent (putting the same configuration does
+  not change anything). Modified configurations must also increase the
+  generation number.
+* `artifact_list`: List the artifacts present in the artifact
+  configuration along with the count of available copies of each
+  artifact across the *update* datasets. Also includes the current
+  generation number.
+* `artifact_put`: Put the request body into the artifact store.
+  Rejects the request if the artifact does not belong to the current
+  configuration.
+* `artifact_copy_from_depot`: Sends a request to another Sled Agent (via
+  the *TUF Repo Depot API*; see below) to fetch an artifact. The base
+  URL for the source sled is chosen by the requester. This API responds
+  after a successful HTTP response from the source sled and the copy
+  proceeds asynchronously. Rejects the request if the artifact does not
+  belong to the current configuration.
+
+Sled Agent also spawns another Dropshot API server called the *TUF Repo
+Depot API* which offers one API on the underlay network, intended for
+other Sled Agents:
+
+* `artifact_get_by_sha256`: Get the content of an artifact.
+
+In an asynchronous task called the _delete reconciler_, Sled Agent
```

Comment:
What happens if a delete occurs while another sled is trying to download the artifact? The other sled may not yet have learned from nexus that the artifact should be deleted because it s not in the present configuration.

Reply:
This is usual POSIX filesystem semantics. If the file is deleted before the requesting sled agent requests it, it will get a 404. If the file is deleted after the sending sled agent opens it, the directory entry is removed, but the file remains open and accessible until the sending sled agent closes the file. In any case if a file is truncated or corrupted in transit, the receiving sled agent will refuse to write it to the artifact store because it verifies the checksum.